types.go

```diff
@@ -0,0 +1,160 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tally
+
+import "time"
+
+// Scope is a namespace wrapper around a stats reporter, ensuring that
+// all emitted values have a given prefix or set of tags
+type Scope interface {
+	// Counter returns the Counter object corresponding to the name
+	Counter(name string) Counter
+
+	// Gauge returns the Gauge object corresponding to the name
+	Gauge(name string) Gauge
+
+	// Timer returns the Timer object corresponding to the name
+	Timer(name string) Timer
+
+	// Tagged returns a new scope with the given tags
+	Tagged(tags map[string]string) Scope
+
+	// SubScope returns a new child scope with the given name
+	SubScope(name string) Scope
+
+	// Capabilities returns a description of metrics reporting capabilities
+	Capabilities() Capabilities
+
+	// Snapshot returns a copy of all values since the last report execution,
+	// this is an expensive operation and should only be use for testing purposes
+	Snapshot() Snapshot
+}
+
+// RootScope is a scope that manages itself and other Scopes
+type RootScope interface {
+	Scope
+
+	// Close Ceases periodic reporting of the root scope and subscopes
+	Close()
+}
+
+// Capabilities is a description of metrics reporting capabilities
+type Capabilities interface {
+	// Reporting returns whether the reporter has the ability to actively report
+	Reporting() bool
+
+	// Tagging returns whether the reporter has the capability for tagged metrics
+	Tagging() bool
+}
+
+// Snapshot is a snapshot of values since last report execution
+type Snapshot interface {
+	// Counters returns a snapshot of all counters since last report execution
+	Counters() map[string]CounterSnapshot
+
+	// Gauges returns a snapshot of all counters since last report execution
+	Gauges() map[string]GaugeSnapshot
+
+	// Timers returns a snapshot of all counters since last report execution
+	Timers() map[string]TimerSnapshot
+}
+
+// CounterSnapshot is a snapshot of a counter
+type CounterSnapshot interface {
+	// Name returns the name
+	Name() string
+
+	// Tags returns the tags
+	Tags() map[string]string
+
+	// Value returns the value
+	Value() int64
+}
+
+// GaugeSnapshot is a snapshot of a counter
+type GaugeSnapshot interface {
+	// Name returns the name
+	Name() string
+
+	// Tags returns the tags
+	Tags() map[string]string
+
+	// Value returns the value
+	Value() int64
+}
+
+// TimerSnapshot is a snapshot of a counter
+type TimerSnapshot interface {
+	// Name returns the name
+	Name() string
+
+	// Tags returns the tags
+	Tags() map[string]string
+
+	// Values returns the values
+	Values() []time.Duration
+}
+
+// StatsReporter is a backend for Scopes to report metrics to
+type StatsReporter interface {
```

Comment:
I would prefer to keep all public APIs in single file for a project of this size.